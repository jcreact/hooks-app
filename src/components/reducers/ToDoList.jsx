import React from 'react';
import PropTypes from 'prop-types';

//Componentes
import { ToDoListItem } from './ToDoListItem.jsx';

/**
 * Renderiza el listado de tareas.
 * @param {Object} props - Propiedades del componente.
 * @param {{id: number, project:string, task:string, done:boolean}[]} props.todos - Lista de tareas.
 * @param {(id: number) => void} props.removeTask - Referencia a la función para remover una tarea.
 * @param {(id: number) => void} props.handleToggle - Referencia a la función para cambiar el estado de la tarea.
 */
export const ToDoList = ({ todos, removeTask, handleToggle }) => {
    const showToDos = () =>
        todos.map((todo, i) => (
            <ToDoListItem
                key={todo.id}
                todo={todo}
                index={i}
                removeTask={removeTask}
                handleToggle={handleToggle}
            />
        ));

    return <ol className="list-group">{showToDos()}</ol>;
};

// Especificación de tipos de las propiedades.
ToDoList.propTypes = {
    todos: PropTypes.arrayOf(
        PropTypes.exact({
            id: PropTypes.number.isRequired,
            project: PropTypes.string.isRequired,
            task: PropTypes.string.isRequired,
            done: PropTypes.bool.isRequired,
        })
    ),
    removeTask: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
};
