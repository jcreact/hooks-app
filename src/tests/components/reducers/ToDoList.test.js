

import { shallow } from 'enzyme';

// Componente
import { ToDoList } from '../../../components/reducers/ToDoList.jsx';

// Datos de prueba
import { demoTodos } from '../../fixtures/demoTodos';

describe(`Pruebas en el componente 'ToDoList.jsx'.`, () => {


    const removeTask = jest.fn();
    const handleToggle = jest.fn();
    const component = shallow(
        <ToDoList
            todos={demoTodos}
            handleToggle={handleToggle}
            removeTask={removeTask}
        />
    );

    test(`Debe renderizar el componente ToDoListItem correctamente.`, () => {
        expect(component).toMatchSnapshot();
        expect(component.find('ToDoListItem')).toHaveLength(demoTodos.length);
    });


    test(`Debe renderizar el listado de tareas correctamente.`, () => {
        expect(component.find('ToDoListItem')).toHaveLength(demoTodos.length);
    });

    test(`Debe de establecer las propieades del componente ToDoListItem.`, () => {

        const index = 0;
        const item = component.find('ToDoListItem').at(index);

        expect(item.prop('handleToggle')).toEqual(expect.any(Function));
        expect(item.prop('removeTask')).toEqual(expect.any(Function));
        expect(item.prop('index')).toBe(index);
        expect(item.prop('todo')).toEqual(demoTodos[index]);

    });

});
