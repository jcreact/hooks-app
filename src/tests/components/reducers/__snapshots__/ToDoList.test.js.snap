// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente 'ToDoList.jsx'. Debe renderizar el componente ToDoListItem correctamente. 1`] = `
<ol
  className="list-group"
>
  <ToDoListItem
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    removeTask={[MockFunction]}
    todo={
      Object {
        "done": false,
        "id": 1,
        "project": "Personal",
        "task": "Tarea 1",
      }
    }
  />
  <ToDoListItem
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    removeTask={[MockFunction]}
    todo={
      Object {
        "done": true,
        "id": 2,
        "project": "Personal",
        "task": "Tarea 2",
      }
    }
  />
  <ToDoListItem
    handleToggle={[MockFunction]}
    index={2}
    key="3"
    removeTask={[MockFunction]}
    todo={
      Object {
        "done": false,
        "id": 3,
        "project": "Personal",
        "task": "Tarea 3",
      }
    }
  />
</ol>
`;
